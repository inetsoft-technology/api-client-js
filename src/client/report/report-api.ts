/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Report contains the properties of a report.
 */
export interface Report {
   /**
    * The path to the report.
    */
   path: string;

   /**
    * The display name of the report.
    */
   label?: string;

   /**
    * The runtime identifier of the opened report.
    */
   identifier?: string;
}

/**
 * ReportParameter contains the definition of a report initialization or creation parameter.
 */
export interface ReportParameter {
   /**
    * The type of the parameter. The allowable values are 'INITIALIZE' and 'CREATE'.
    */
   type: string;

   /**
    * The data type of the parameter value. The allowable values are 'string', 'boolean', 'float',
    * 'double', 'decimal', 'char', 'character', 'byte', 'short', 'integer', 'long', 'timeInstant',
    * 'date', and 'time'.
    */
   dataType: string;

   /**
    * The name of the parameter.
    */
   name: string;

   /**
    * The value of the parameter.
    */
   value: string;
}

/**
 * PermissionGrant contains a permission granted to an identity.
 */
export interface PermissionGrant {
   /**
    * The the name of the identity.
    */
   name: string;

   /**
    * The type of the identity. The allowable values are 'USER', 'GROUP', and 'ROLE'.
    */
   type: string;

   /**
    * The list of granted actions. The allowed values are 'READ', 'WRITE', 'DELETE', 'ACCESS',
    * 'ASSIGN', and 'ADMIN'.
    */
   actions: string[];
}

/**
 * ReportApi provides operations used to execute reports.
 */
export interface ReportApi {
   /**
    * Gets the list of reports.
    */
   getReports(): Promise<Report[]>;

   /**
    * Gets the list of opened reports.
    */
   getOpenReports(): Promise<Report[]>;

   /**
    * Opens a report.
    *
    * @param path       the path to the report to open.
    * @param execute    true to immediately execute the report; false otherwise. If false, a
    *                   subsequent call to executeReport() is required.
    * @param parameters the parameters to use when opening the report.
    *
    * @return the opened report.
    */
   openReport(path: string, execute: boolean, parameters: {[name: string]: any}): Promise<Report>;

   /**
    * Gets information about an opened report.
    *
    * @param id the identifier of the opened report.
    *
    * @return the opened report.
    */
   getOpenedReport(id: string): Promise<Report>;

   /**
    * Executes an opened report.
    *
    * @param id         the identifier of the opened report.
    * @param parameters the parameters to use when executing the report.
    */
   executeReport(id: string, parameters: {[name: string]: any}): Promise<void>;

   /**
    * Closes an opened report.
    *
    * @param id the identifier of the opened report.
    */
   closeReport(id: string): Promise<void>;

   /**
    * Gets the number of pages in a report.
    *
    * @param id the identifier of the opened report.
    *
    * @return the number of pages in the specified report.
    */
   getReportPageCount(id: string): Promise<number>;

   /**
    * Sends an opened report in an email message.
    *
    * @param id         the identifier of the opened report.
    * @param recipients the email addresses of the recipients of the message.
    * @param sender     the email address of the sender of the message.
    * @param subject    the subject line of the message.
    * @param message    the body of the email message.
    * @param format     the file format for the attached report.
    */
   mailReport(id: string, recipients: string[], sender: string, subject: string, message: string,
              format: string): Promise<void>;

   /**
    * Prints an opened report on the server.
    *
    * @param id      the identifier of the opened report.
    * @param printer the name of the server printer.
    * @param pages   the page numbers to print.
    */
   printReport(id: string, printer: string, pages: number[]): Promise<void>;

   /**
    * Exports an opened report.
    *
    * @param id       the identifier of the opened report.
    * @param format   the file format for the exported report.
    *
    * @return the binary content of the exported report.
    */
   exportReport(id: string, format: string): Promise<ArrayBuffer>;

   /**
    * Deploys a report to the repository.
    *
    * @param file          the binary content of the template file.
    * @param name          the name of the deployed report.
    * @param alias         the alias for the report.
    * @param description   a description of the report.
    * @param folder        the path to the parent folder of the report.
    * @param visible       a flag that indicates if the report is visible in the repository tree.
    * @param bursting      a flag that indicates if bursting is enabled for the report.
    * @param pregenerated  a flag that indicates if the report is pre-generated.
    * @param cycle         the name of the cycle to be used to pre-generate the report.
    * @param overwrite     a flag that indicates if an existing report should be overwritten.
    * @param exportFormats the list of export formats that are allowed for the report.
    * @param parameters    the initialization and creation parameters for the report.
    * @param permissions   the permissions granted on the report.
    */
   deployReport(file: ArrayBuffer, name: string, alias: string, description: string, folder: string,
                visible: boolean, bursting: boolean, pregenerated: boolean, cycle: string | null,
                overwrite: boolean, exportFormats: string[], parameters: ReportParameter[],
                permissions: PermissionGrant[]): Promise<void>;
}
