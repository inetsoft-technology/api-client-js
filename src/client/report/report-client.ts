/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { fromByteArray } from "base64-js";
import { AbstractClient } from "../abstract-client";
import { HttpAdapter } from "../http-adapter";
import { PermissionGrant, Report, ReportApi, ReportParameter } from "./report-api";

/**
 * @hidden
 */
export class ReportClient extends AbstractClient implements ReportApi {

   constructor(adapter: HttpAdapter<any>, token: () => Promise<string>, baseUrl: string) {
      super(adapter, token, baseUrl);
   }

   public getReports(): Promise<Report[]> {
      const url = "api/public/reports";
      return this.doGet(url).then((list: any) => list.reports);
   }

   public getOpenReports(): Promise<Report[]> {
      const url = "api/public/reports/open";
      return this.doGet(url).then((list: any) => list.reports);
   }

   public openReport(path: string, execute: boolean,
                     parameters: {[name: string]: any}): Promise<Report>
   {
      const request = { path, execute, parameters };
      const url = "api/public/reports/open";
      return this.doPost(url, request);
   }

   public getOpenedReport(id: string): Promise<Report> {
      const url = `api/public/reports/open/${id}`;
      return this.doGet(url);
   }

   public executeReport(id: string, parameters: {[name: string]: any}): Promise<void> {
      const request = { parameters };
      const url = `api/public/reports/open/${id}/execute`;
      return this.doPost(url, request);
   }

   public closeReport(id: string): Promise<void> {
      const url = `api/public/reports/open/${id}`;
      return this.doDelete(url);
   }

   public getReportPageCount(id: string): Promise<number> {
      const url = `api/public/reports/open/${id}/pages`;
      return this.doGet(url).then((count: any) => count.pages);
   }

   public mailReport(id: string, recipients: string[], sender: string, subject: string,
                     message: string, format: string): Promise<void>
   {
      const request = { recipients, sender, subject, message, format };
      const url = `api/public/reports/open/${id}/mail`;
      return this.doPost(url, request);
   }

   public printReport(id: string, printer: string, pages: number[]): Promise<void> {
      const request = { printer, pages };
      const url = `api/public/reports/open/${id}/print`;
      return this.doPost(url, request);
   }

   public exportReport(id: string, format: string): Promise<ArrayBuffer> {
      const url = `api/public/reports/open/${id}/export/${format}`;
      return this.doGetBinary(url);
   }

   public deployReport(file: ArrayBuffer, name: string, alias: string, description: string,
                       folder: string, visible: boolean, bursting: boolean, pregenerated: boolean,
                       cycle: string | null, overwrite: boolean, exportFormats: string[],
                       parameters: ReportParameter[], permissions: PermissionGrant[]): Promise<void>
   {
      const request = {
         data: fromByteArray(new Uint8Array(file)), name, alias, description, folder, visible,
         bursting, pregenerated, cycle, overwrite, exportFormats, parameters, permissions
      };
      const url = "api/public/reports/deploy";
      return this.doPost(url, request);
   }
}
