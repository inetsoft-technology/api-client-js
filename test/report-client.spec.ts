/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ReportClient } from "../src/client/report/report-client";

describe("ReportClient", () => {
   let adapter: any;
   let client: ReportClient;

   const reports = [
      {
         path: "folder/report1",
         label: "Folder/Report 1"
      },
      {
         path: "folder/report2",
         label: "Folder/Report 2"
      },
      {
         path: "folder/report3",
         label: "Folder/Report 3"
      }
   ];

   const openReports = [
      {
         path: "folder/report1",
         label: "Folder/Report 1",
         identifier: "id1"
      },
      {
         path: "folder/report2",
         label: "Folder/Report 2",
         identifier: "id2"
      }
   ];

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn(),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new ReportClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when getReports() is called", () => {
      it("then a list of reports is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {reports}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getReports().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports");
         });
      });
   });

   describe("when getOpenReports() is called", () => {
      it("then a list of reports is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {reports: openReports}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getOpenReports().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open");
         });
      });
   });

   describe("when openReport() is called and execute is true", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({data: openReports[0]}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.openReport("folder/report1", true, {p1: "v1", p2: 1}).then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when openReport() is called and execute is false", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({data: openReports[0]}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.openReport("folder/report1", false, {p1: "v1", p2: 1}).then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when getOpenedReport() is called", () => {
      it("then the opened report is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: openReports[0]}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getOpenedReport("id1").then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1");
         });
      });
   });

   describe("when executeReport() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.executeReport("id1", {p1: "v1", p2: 1}).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1/execute");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when closeReport() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.delete.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.closeReport("id1").then(() => {
            expect(adapter.delete.mock.calls.length).toBe(1);
            expect(adapter.delete.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1");
         });
      });
   });

   describe("when getReportPageCount() is called", () => {
      it("then the number of pages is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {pages: 5}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getReportPageCount("id1").then(result => {
            expect(result).toBe(5);
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1/pages");
         });
      });
   });

   describe("when mailReport() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.mailReport("id1", ["joe@example.com", "sue@example.com"], "ann@example.com", "Test Mail", "This is a test email.", "PDF").then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1/mail");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when printReport() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.printReport("id1", "Test Printer", [1, 2, 3, 4]).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1/print");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when exportReport() is called", () => {
      it("then the exported report file is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: createBinaryData()}));
         adapter.mapBinary.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.exportReport("id1", "PDF").then(result => {
            validateBinaryData(result);
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/open/id1/export/PDF");
         });
      });
   });

   describe("when deployReport() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         const parameters = [{
            type: "CREATE",
            dataType: "string",
            name: "param1",
            value: "value1"
         }];
         const permissions = [{
            name: "admin",
            type: "USER",
            actions: ["READ", "WRITE", "DELETE", "ADMIN"]
         }];

         return client.deployReport(createBinaryData(), "report4", "Report 4", "A test report", "folder2", true, false, false, null, true, ["PDF", "EXCEL"], parameters, permissions).then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/reports/deploy");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });
});

function createBinaryData(): ArrayBuffer {
   const buffer = new ArrayBuffer(16);
   const array = new Int32Array(buffer);
   array[0] = 1;
   array[1] = 2;
   array[2] = 3;
   array[3] = 4;
   return buffer;
}

function validateBinaryData(buffer: ArrayBuffer): void {
   expect(buffer).toBeTruthy();
   expect(buffer.byteLength).toBe(16);
   const actual = new Int32Array(buffer);
   expect(actual[0]).toBe(1);
   expect(actual[1]).toBe(2);
   expect(actual[2]).toBe(3);
   expect(actual[3]).toBe(4);
}
