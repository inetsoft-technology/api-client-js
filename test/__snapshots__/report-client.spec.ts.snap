// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportClient when deployReport() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "alias": "Report 4",
  "bursting": false,
  "cycle": null,
  "data": "AQAAAAIAAAADAAAABAAAAA==",
  "description": "A test report",
  "exportFormats": Array [
    "PDF",
    "EXCEL",
  ],
  "folder": "folder2",
  "name": "report4",
  "overwrite": true,
  "parameters": Array [
    Object {
      "dataType": "string",
      "name": "param1",
      "type": "CREATE",
      "value": "value1",
    },
  ],
  "permissions": Array [
    Object {
      "actions": Array [
        "READ",
        "WRITE",
        "DELETE",
        "ADMIN",
      ],
      "name": "admin",
      "type": "USER",
    },
  ],
  "pregenerated": false,
  "visible": true,
}
`;

exports[`ReportClient when executeReport() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "parameters": Object {
    "p1": "v1",
    "p2": 1,
  },
}
`;

exports[`ReportClient when getOpenReports() is called then a list of reports is returned 1`] = `
Array [
  Object {
    "identifier": "id1",
    "label": "Folder/Report 1",
    "path": "folder/report1",
  },
  Object {
    "identifier": "id2",
    "label": "Folder/Report 2",
    "path": "folder/report2",
  },
]
`;

exports[`ReportClient when getOpenedReport() is called then the opened report is returned 1`] = `
Object {
  "identifier": "id1",
  "label": "Folder/Report 1",
  "path": "folder/report1",
}
`;

exports[`ReportClient when getReports() is called then a list of reports is returned 1`] = `
Array [
  Object {
    "label": "Folder/Report 1",
    "path": "folder/report1",
  },
  Object {
    "label": "Folder/Report 2",
    "path": "folder/report2",
  },
  Object {
    "label": "Folder/Report 3",
    "path": "folder/report3",
  },
]
`;

exports[`ReportClient when mailReport() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "format": "PDF",
  "message": "This is a test email.",
  "recipients": Array [
    "joe@example.com",
    "sue@example.com",
  ],
  "sender": "ann@example.com",
  "subject": "Test Mail",
}
`;

exports[`ReportClient when openReport() is called and execute is false then the adapter is called with the correct arguments 1`] = `
Object {
  "identifier": "id1",
  "label": "Folder/Report 1",
  "path": "folder/report1",
}
`;

exports[`ReportClient when openReport() is called and execute is false then the adapter is called with the correct arguments 2`] = `
Object {
  "execute": false,
  "parameters": Object {
    "p1": "v1",
    "p2": 1,
  },
  "path": "folder/report1",
}
`;

exports[`ReportClient when openReport() is called and execute is true then the adapter is called with the correct arguments 1`] = `
Object {
  "identifier": "id1",
  "label": "Folder/Report 1",
  "path": "folder/report1",
}
`;

exports[`ReportClient when openReport() is called and execute is true then the adapter is called with the correct arguments 2`] = `
Object {
  "execute": true,
  "parameters": Object {
    "p1": "v1",
    "p2": 1,
  },
  "path": "folder/report1",
}
`;

exports[`ReportClient when printReport() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "pages": Array [
    1,
    2,
    3,
    4,
  ],
  "printer": "Test Printer",
}
`;
